### DISCUSSION TITLE
@mdit-vue/plugin-component tries several ways to render a vue component, but it doesn't work.

### DISCUSSION THREAD

**Discussion by maquannene** [mdit-vue/mdit-vue#8] [Q&A] [answered]
I wonder if there is something wrong with my understanding of how to use this library.

I registered the Foo Vue3 component globally, and markdown content is also used <Foo />,  But Foo is still not render.

Could you please give me a specific example, or give me some ideas about how to use it?

**Comment by brc-dd** [comment, marked as answer]
Can you show what you've tried? (in form of some GitHub repo or StackBlitz link)

You probably should use some wrapper like https://github.com/unplugin/unplugin-vue-markdown instead if you're not comfortable with low-level markdown-it packages in this repo.

**Reply by maquannene** [reply]
ohhhhhhhhhh, I understand, this package is only used for parsing, and can not provide the actual render, the render part still have to be implemented by myself.

I took a look at unplugin-vue-markdown， and it should meet my needs. Thank you very much.

**Reply by maquannene** [reply]
I'm sorry to bother you again.
I saw this package https://github.com/unplugin/unplugin-vue-markdown, may still not meet my requirements, as this [issue](https://github.com/unplugin/unplugin-vue-markdown/issues/28) description, use this package only supports the compile time, does not support runtime, And it's not a plugin for [markdown-it](https://github.com/markdown-it/markdown-it).

So I still need to implement the function of rendering vue3 component in cotent of markdown at runtime by myself. Do you have any good suggestions or methods?

**Reply by brc-dd** [reply]
mdit-vue plugins can help you convert your markdown into vue sfc. But you'll need to use something to convert the string to actual vue component.

**Reply by maquannene** [reply]
Sorry, I mean I want to embed Vue component into markdown instead of convert markdown into Vue render.

Markdown source content like this：

```
// md
# Title
<VueComponent  id="123" />
```

// finally render
Title 
VueComponent（implement by vue3 or other UI Framework，the component is relatively complicated）


So I originally thought that this plugin（@mdit-vue/plugin-component） would allow me to write <VueComponent  id="123" />  directly in the markdown source content, and during execution it would help me change <VueComponent> tag to use vue 
 component rendering.

**Reply by brc-dd** [reply]
You'll need both @mdit-vue/plugin-component and @mdit-vue/plugin-sfc. At the end it should be a vue component that can be rendered by vue.

**Reply by maquannene** [reply]
I really couldn't understand the usage of these two packages, so I gave up.

**Reply by meteorlxy** [reply]
> and during execution it would help me change tag to use vue

What do you mean by "during execution"?

markdown-it can only help you to transform markdown to html. Notice that the vue tags will be kept as is during the transformation, because markdown-it could not "execute" vue components. They will be treated as normal HTML tags. Also, mdit-vue plugins could NOT help to "execute" them.

Anyway, if you could not understand the process, you should not use these packages for sure. These mdit-vue plugins are mainly provided for library authors.

**Reply by brc-dd** [reply]
I think OP is trying to build something like the editor+preview in https://vue-styleguidist.github.io/docgen-vite/components/HelloWorld.html (with support for vue in markdown syntax)

One can get sfc as string by something like this:

```ts
import { componentPlugin } from '@mdit-vue/plugin-component'
import { sfcPlugin } from '@mdit-vue/plugin-sfc'
import type { MarkdownItEnv } from '@mdit-vue/types'
import MarkdownIt from 'markdown-it'

let md: MarkdownIt | null = null

export function useMarkdownRenderer() {
  if (!md) {
    md = new MarkdownIt({ html: true }).use(componentPlugin).use(sfcPlugin)
  }

  return renderMarkdown
}

function renderMarkdown(content: string) {
  const env: MarkdownItEnv = {}
  md!.render(content, env)

  return [
    ...(env.sfcBlocks?.scripts.map((block) => block.content) ?? []),
    env.sfcBlocks?.template?.content,
    ...(env.sfcBlocks?.styles.map((block) => block.content) ?? []),
    ...(env.sfcBlocks?.customBlocks.map((block) => block.content) ?? [])
  ]
    .filter(Boolean)
    .join('\n')
}

// usage:

const mdToVue = useMarkdownRenderer()

const mdSource = '...'

const vueSource = mdToVue(mdSource) // will be string not Component

// do something with vueSource like feed to styleguidist or vue3-sfc-loader
// vueSource will be something like this for the example OP provided:
// `<template><h1>Title</h1>\n<VueComponent  id="123" /></template>`
```
